### ISSUE TITLE
Question about sim2sim.py - Robot "Takes Off" with Custom 20-DOF Model

### ISSUE THREAD

**Issue by wy12184163** [Open-X-Humanoid/TienKung-Lab#10]
Hello,
First, thank you for providing such a great framework for learning. I have successfully trained my custom 20-DOF robot, and it achieves good performance within Isaac Lab.
I am now trying to test my robot using the sim2sim.py script, but I'm running into an issue where the robot "takes off" (flies up uncontrollably) as soon as control is enabled. I can confirm that without any control commands, the robot loads correctly and remains stable. This leads me to believe the issue is with the sim-to-sim interface.
I have a couple of points of confusion regarding the sim-to-sim setup:
1. Regarding the two joint index mapping arrays:
My understanding of self.mujoco_to_isaac_idx and self.isaac_to_mujoco_idx is as follows:
self.mujoco_to_isaac_idx: This array should be ordered according to the joint order as read by Isaac Lab. The values in this array should be the corresponding actuator indices from the MuJoCo model.
self.isaac_to_mujoco_idx: Conversely, this array should be ordered according to the actuator order in the MuJoCo model. The values in this array should be the corresponding joint indices as read by Isaac Lab.
self.mujoco_to_isaac_idx = [
            0,  # hip_roll_l_joint
            6,  # hip_roll_r_joint
            12,  # shoulder_pitch_l_joint
            16,  # shoulder_pitch_r_joint
            1,  # hip_pitch_l_joint
            7,  # hip_pitch_r_joint
            13,  # shoulder_roll_l_joint
            17,  # shoulder_roll_r_joint
            2,  # hip_yaw_l_joint
            8,  # hip_yaw_r_joint
            14,  # shoulder_yaw_l_joint
            18,  # shoulder_yaw_r_joint
            3,  # knee_pitch_l_joint
            9,  # knee_pitch_r_joint
            15,  # elbow_pitch_l_joint
            19,  # elbow_pitch_r_joint
            4,  # ankle_pitch_l_joint
            10,  # ankle_pitch_r_joint
            5,  # ankle_roll_l_joint
            11,  # ankle_roll_r_joint
        ]
        self.isaac_to_mujoco_idx = [
            0,  # hip_roll_l_joint
            4,  # hip_pitch_l_joint
            8,  # hip_yaw_l_joint
            12,  # knee_pitch_l_joint
            16,  # ankle_pitch_l_joint
            18,  # ankle_roll_l_joint
            1,  # hip_roll_r_joint
            5,  # hip_pitch_r_joint
            9,  # hip_yaw_r_joint
            13,  # knee_pitch_r_joint
            17,  # ankle_pitch_r_joint
            19,  # ankle_roll_r_joint
            2,  # shoulder_pitch_l_joint
            6,  # shoulder_roll_l_joint
            10,  # shoulder_yaw_l_joint
            14,  # elbow_pitch_l_joint
            3,  # shoulder_pitch_r_joint
            7,  # shoulder_roll_r_joint
            11,  # shoulder_yaw_r_joint
            15,  # elbow_pitch_r_joint
        ]
After modifying the mappings based on this understanding, the robot is still not controlled correctly. Is my understanding of these indexing arrays correct?
2. Regarding the choice of control mode:
I have noticed that many previous sim-to-sim implementations use PD control for testing. However, this framework uses position control. Is there a specific reason for this design choice? We also tried changing the actuators in our MuJoCo XML to use position control, but this did not resolve the issue.
These are my main points of confusion. Any help or clarification would be greatly appreciated.
Thank you

**Comment by mggeng** [comment]
I think the code shown up is correct, so i think maybe it is the joint order that is not right. Maybe you could debug or print while running play scripts and compare the action value with mujoco.

**Comment by BryceWang0527** [comment]
Hi, glad to hear your training went well.

1. The index mismatch comes from the different default orderings in Isaac Sim and MuJoCo. 
- In Isaac Sim, print **self.robot.data.joint_names**; 
- in MuJoCo, use the order defined by the actuators in the MJCF. Align the two lists as in the original repo.

2. We use position control in sim2sim because training calls self.robot.set_joint_position_target(). Torque control also works if you train with it—just keep the same mode.

**Comment by Serissa** [comment]
> Hello, First, thank you for providing such a great framework for learning. I have successfully trained my custom 20-DOF robot, and it achieves good performance within Isaac Lab. I am now trying to test my robot using the sim2sim.py script, but I'm running into an issue where the robot "takes off" (flies up uncontrollably) as soon as control is enabled. I can confirm that without any control commands, the robot loads correctly and remains stable. This leads me to believe the issue is with the sim-to-sim interface. I have a couple of points of confusion regarding the sim-to-sim setup:
> 
>     1. Regarding the two joint index mapping arrays:
>        My understanding of self.mujoco_to_isaac_idx and self.isaac_to_mujoco_idx is as follows:
>        self.mujoco_to_isaac_idx: This array should be ordered according to the joint order as read by Isaac Lab. The values in this array should be the corresponding actuator indices from the MuJoCo model.
>        self.isaac_to_mujoco_idx: Conversely, this array should be ordered according to the actuator order in the MuJoCo model. The values in this array should be the corresponding joint indices as read by Isaac Lab.
>        self.mujoco_to_isaac_idx = [
>        0,  # hip_roll_l_joint
>        6,  # hip_roll_r_joint
>        12,  # shoulder_pitch_l_joint
>        16,  # shoulder_pitch_r_joint
>        1,  # hip_pitch_l_joint
>        7,  # hip_pitch_r_joint
>        13,  # shoulder_roll_l_joint
>        17,  # shoulder_roll_r_joint
>        2,  # hip_yaw_l_joint
>        8,  # hip_yaw_r_joint
>        14,  # shoulder_yaw_l_joint
>        18,  # shoulder_yaw_r_joint
>        3,  # knee_pitch_l_joint
>        9,  # knee_pitch_r_joint
>        15,  # elbow_pitch_l_joint
>        19,  # elbow_pitch_r_joint
>        4,  # ankle_pitch_l_joint
>        10,  # ankle_pitch_r_joint
>        5,  # ankle_roll_l_joint
>        11,  # ankle_roll_r_joint
>        ]
>        self.isaac_to_mujoco_idx = [
>        0,  # hip_roll_l_joint
>        4,  # hip_pitch_l_joint
>        8,  # hip_yaw_l_joint
>        12,  # knee_pitch_l_joint
>        16,  # ankle_pitch_l_joint
>        18,  # ankle_roll_l_joint
>        1,  # hip_roll_r_joint
>        5,  # hip_pitch_r_joint
>        9,  # hip_yaw_r_joint
>        13,  # knee_pitch_r_joint
>        17,  # ankle_pitch_r_joint
>        19,  # ankle_roll_r_joint
>        2,  # shoulder_pitch_l_joint
>        6,  # shoulder_roll_l_joint
>        10,  # shoulder_yaw_l_joint
>        14,  # elbow_pitch_l_joint
>        3,  # shoulder_pitch_r_joint
>        7,  # shoulder_roll_r_joint
>        11,  # shoulder_yaw_r_joint
>        15,  # elbow_pitch_r_joint
>        ]
>        After modifying the mappings based on this understanding, the robot is still not controlled correctly. Is my understanding of these indexing arrays correct?
> 
>     2. Regarding the choice of control mode:
>        I have noticed that many previous sim-to-sim implementations use PD control for testing. However, this framework uses position control. Is there a specific reason for this design choice? We also tried changing the actuators in our MuJoCo XML to use position control, but this did not resolve the issue.
>        These are my main points of confusion. Any help or clarification would be greatly appreciated.
>        Thank you

1. Replace the urdf of tienkung2 with that of Yushu g1, generate the data of g1 in the tienkung2 data format, and verify the data in motion_amp_expert and motion_visualization.Modify the offset, speed limit and other values in tienkung.py according to g1.
2. It was found that the output of self.left_leg_ids in tienkung_env.py is [0, 4, 8, 12, 16, 18], and the output of self.right_leg_ids is [1, 5, 9, 13, 17, 19].Compared with the output of g1, the output of self.left_leg_ids is [4, 0, 8, 12, 16, 18], and that of self.right_leg_ids is [5, 1, 9, 13, 17, 19].This should be caused by the different sequences of hip_roll and hip_pitch in tienkung2 and g1.
3. Modify def hip_roll_action(env: TienKungEnv/G1Env) -> torch.Tensor:
# return torch.sum(torch.abs(env.action[:, [env.left_leg_ids[0], env.right_leg_ids[0]]]), dim=1)
return torch.sum(torch.abs(env.action[:, [env.left_leg_ids[1], env.right_leg_ids[1]]]), dim=1)
4. After training to 10,000 steps, the values of Mean amp_policy_pred loss and Mean amp_expert_pred loss oscillate in the attachments of -0.69 and 0.69 respectively.
5. May I ask from which direction I should solve the problem?Thank you very much for your guidance!
